server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -0,0 +1,403 @@
+
+        this.desiredBalanceComputation = new ContinuousComputation<>(threadPool.generic()) {
+            @Override
+            protected void processInput(RerouteInput actualRoutingAllocation) {
+                updateDesiredBalanceAndReroute(actualRoutingAllocation);
+            }
+        };
+    }
+
+    @Override
+    public void allocate(RoutingAllocation allocation) {
+        assert MasterService.isMasterUpdateThread() || Thread.currentThread().getName().startsWith("TEST-")
+            : Thread.currentThread().getName();
+        // assert allocation.debugDecision() == false; set to true when called via the reroute API
+        assert allocation.ignoreDisable() == false;
+
+        desiredBalanceComputation.onNewInput(
+            new RerouteInput(allocation.immutableClone(), new ArrayList<>(allocation.routingNodes().unassigned().ignored()))
+        );
+
+        new Reconciler(currentDesiredBalance, allocation).run();
+
+    }
+
+    @Override
+    public ShardAllocationDecision decideShardAllocation(ShardRouting shard, RoutingAllocation allocation) {
+        return balancedShardsAllocator.decideShardAllocation(shard, allocation);
+    }
+
+    /**
+     * @param routingAllocation a copy of the immutable parts of the allocation decision process context
+     * @param ignoredShards     a copy of the shards for which earlier allocators have claimed responsibility
+     */
+    record RerouteInput(RoutingAllocation routingAllocation, List<ShardRouting> ignoredShards) {}
+
+    private void updateDesiredBalanceAndReroute(RerouteInput rerouteInput) {
+
+        final var routingAllocation = rerouteInput.routingAllocation().mutableCloneForSimulation();
+        final var routingNodes = routingAllocation.routingNodes();
+        final var ignoredShards = new HashSet<>(rerouteInput.ignoredShards());
+        final var desiredBalance = currentDesiredBalance;
+        final var changes = routingAllocation.changes();
+        final var knownNodeIds = routingAllocation.nodes().stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+
+        // we assume that all ongoing recoveries will complete
+        for (final var routingNode : routingNodes) {
+            for (final var shardRouting : routingNode) {
+                if (shardRouting.initializing()) {
+                    routingNodes.startShard(logger, shardRouting, changes);
+                    // TODO adjust disk usage info to reflect the assumed shard movement
+                }
+            }
+        }
+
+        // we are not responsible for allocating unassigned primaries of existing shards, and we're only responsible for allocating
+        // unassigned replicas if the ReplicaShardAllocator gives up, so we must respect these ignored shards
+        final RoutingNodes.UnassignedShards unassigned = routingNodes.unassigned();
+        for (final var shardRouting : unassigned) {
+            if (ignoredShards.contains(shardRouting)) {
```

Comment:
The existing-shards allocators call `removeAndIgnore` on a different `RoutingNodes`. We create our own `RoutingNodes` in `mutableCloneForSimulation` so we need to do this so that we ignore the right shards.

However this does make me realise that we're not seeing any shards that the existing-shards allocators actually allocated in this pass, because they won't have been copied over to the `RoutingTable` yet.